cert/spec/runner_spec.rb

```diff
@@ -73,5 +75,58 @@
       Cert.config = FastlaneCore::Configuration.create(Cert::Options.available_options, keychain_path: ".")
       Cert::Runner.new.revoke_expired_certs!
     end
+
+    describe ":filename option handling" do
+      filename = ""
```

Comment:
This all looks muuuccchhhh better 🙌 The one last thing I would recommend is staying away from putting a global-ish variable like this that gets mutated by the tests. I would probably attempt the following...

```rb
let (:temp) { Dir.tmpdir }
let (:certificate) { stub_certificate }
before do
  allow(Spaceship).to receive(:login).and_return(nil)
  allow(Spaceship).to receive(:client).and_return("client")
  allow(Spaceship).to receive(:select_team).and_return(nil)
  allow(Spaceship.client).to receive(:in_house?).and_return(false)
  allow(Spaceship.certificate.production).to receive(:all).and_return([certificate])
  allow(FastlaneCore::CertChecker).to receive(:installed?).and_return(true)
end

def make_filepath(filename)
  filename_ext = File.extname(filename) == ".cer" ? filename : "#{filename}.cer"
  File.extname(filename) == ".cer" ? "#{temp}/#{filename}" : "#{temp}/#{filename}.cer"
end

context 'with filename flag' do
  it "can forget the file extension" do
    filename = "cert_name"
    filepath = make_filepath(filename)
    expect(File.exist?(filepath)).to be false

    options = { output_path: temp, filename: filename }
    Cert.config = FastlaneCore::Configuration.create(Cert::Options.available_options, options)
    Cert::Runner.new.launch

    expect(File.exist?(filepath)).to be true
    File.delete(filepath)
  end
end
```

This does a few things...
1) `generate` right now are being used as methods instead of variables (they should really be used as variables. By moving the "allows" to a `before` block and moving the `certificate` to a "let", we can get rid of `generate`
2) 'filepath` is also being used as a method. We can actually define this as method and pass a filename to it to generate the `filepath` which allows us to get rid of a global-ish `filename` variable that we are mutating

Let me know if this doesn't work 🙃

Reply:
I am agree with you, but about your first note, you delete `generate` method but it implies to duplicate "Runner" code as many times you need to test, are you sure ? A convenient method named "generate" would do this job : 
```ruby
      let(:generate) do
        options = { output_path: temp, filename: filename }
        Cert.config = FastlaneCore::Configuration.create(Cert::Options.available_options, options)
        Cert::Runner.new.launch
      end

Reply:
@jhanzo I'm okay with a little bit of duplicate runner code 😊 Its okay, IMO, for tests to have some duplicate code. You could create a method instead of doing `let(:generate)` if you want...
```rb
def generate(output_path, filename)
  options = { output_path: temp, filename: filename }
  Cert.config = FastlaneCore::Configuration.create(Cert::Options.available_options, options)
   Cert::Runner.new.launch
end
```
but this isn't totally necessary 😇